### PR TITLE
Tweak: Hide the empty view cell in grid container for some cases [ED-11369}

### PR DESCRIPTION
**Pull request by murzin-nikita**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: assets/dev/js/editor/elements/views/container.js [hunk 1]
```diff
@@ -600,6 +604,10 @@ const ContainerView = BaseElementView.extend( {
 			this.handleGridEmptyView();
 		}
 	},
+
+	isDistortsPreview() {
+		return this.container.children.length === this.container.settings.get( 'grid_columns_grid' ).size * this.container.settings.get( 'grid_rows_grid' ).size;
```
Comment (louiswol94) [inline assets/dev/js/editor/elements/views/container.js:609]: @mykytamurzin Need to handle current device mode etc `grid_rows_grid_tablet` etc. Check how it's done in `getDeviceGridDimensions()`.

File: assets/dev/js/frontend/handlers/container/grid-container.js [hunk 2]
```diff
@@ -273,4 +279,11 @@ export default class GridContainer extends elementorModules.frontend.handlers.Ba
 
 		return ! numberPattern.test( gridRows?.size );
 	}
+
+	shouldRemoveEmptyView() {
+		const elementSettings = this.getElementSettings();
+		const childrenLength = this.elements.outlineParentContainer.querySelectorAll( ':scope > .elementor-element' ).length;
+
+		return childrenLength === elementSettings.grid_columns_grid.size * elementSettings.grid_rows_grid.size;
```
Comment (louiswol94) [inline assets/dev/js/frontend/handlers/container/grid-container.js:287]: Use `getDeviceGridDimensions()`

File: includes/controls/groups/grid-container.php [hunk 3]
```diff
@@ -295,6 +296,17 @@ protected function get_responsive_unit_defaults() {
...
+	protected function get_responsive_autoflow_defaults() {
+		$responsive_autoflow_defaults = [];
+		$active_breakpoints = Plugin::$instance->breakpoints->get_active_breakpoints();
+
+		foreach ( $active_breakpoints as $breakpoint_name => $breakpoint ) {
+			$responsive_autoflow_defaults[ $breakpoint_name . '_default' ] = 'row';
+		}
+
+		return $responsive_autoflow_defaults;
+	}
```
Comment (louiswol94) [inline includes/controls/groups/grid-container.php:308]: @mykytamurzin Why did you need to do this? In the JS you can use `elementorFrontend.utils.controls.getResponsiveControlValue()`. This will get the correct value, if the smaller device is empty it will inherit the value from the bigger device.

Comment (mykytamurzin) [reply includes/controls/groups/grid-container.php:308]: These default values are checking in PHPUnit tests, they were failing without it. See 
https://github.com/elementor/elementor/actions/runs/5781501240/job/15666703826#step:7:177

Comment (mykytamurzin) [reply includes/controls/groups/grid-container.php:308]: I don't need them for my feature, but after adding frontend_available=true - they started popping up

**Review by TzviRabinovitch — DISMISSED** [review]

File: assets/dev/js/frontend/handlers/container/grid-container.js [hunk 4]
```diff
@@ -273,4 +277,56 @@ export default class GridContainer extends elementorModules.frontend.handlers.Ba
...
+		if ( 0 === childrenLength ) {
+			return false;
+		}
+
+		const maxElements = this.getMaxElementsNumber();
+
+		return maxElements <= childrenLength && this.isFullFilled( childrenLength );
+	}
+
+	getMaxElementsNumber() {
```
Comment (davseve) [inline assets/dev/js/frontend/handlers/container/grid-container.js:293]: I would move this function to the end as I am using it later on as well. 
Just for the file order...

File: tests/playwright/sanity/container-grid.test.ts [hunk 5]
```diff
@@ -544,6 +544,47 @@ test.describe( 'Container Grid tests @container', () => {
...
+
+			await editor.removeElement( latestAddedWidgetId );
+		} );
+
+		await test.step( 'Increase number of rows to see empty item', async () => {
+			await editor.selectElement( containerId );
+			await gridRowsControl.locator( '.elementor-slider-input input' ).fill( '3' );
+			await expect( editor.getPreviewFrame().locator( '.elementor-empty-view' ) ).toBeVisible();
+		} );
+	} );
```
Comment (davseve) [inline tests/playwright/sanity/container-grid.test.ts:587]: I would add a test that checks the device mode change.
E,g. On the desktop, the grid is 4*3 and contains 12 items. on the tablet, the grid is 3*3 so in the tablet we expect to see the empty view.

**Review by davseve — APPROVED** [review]
